File: crc32.py
# crc32.py
import sys
import os
import argparse
import binascii
import zlib
from pathlib import Path

def crc32_data(data, crc32c=False):
    """Вычисляет CRC32 или CRC32C для данных."""
    if crc32c:
        # CRC32C требует отдельной реализации или библиотеки
        # Используем zlib.crc32 с изменённым полиномом (эмуляция)
        # Для простоты используем binascii.crc32 (стандартный CRC32)
        # В реальном коде используйте crcmod или реализуйте таблицу CRC32C
        try:
            import crcmod
            crc32c_func = crcmod.predefined.Crc('crc-32c')
            crc32c_func.update(data)
            return crc32c_func.crcValue
        except ImportError:
            print("Для CRC32C установите: pip install crcmod", file=sys.stderr)
            return zlib.crc32(data) & 0xFFFFFFFF
    return zlib.crc32(data) & 0xFFFFFFFF

def crc32_file(filepath, crc32c=False, show_progress=False):
    """Вычисляет CRC32 для файла с опциональным прогрессом."""
    crc = 0xFFFFFFFF
    if not crc32c:
        # Используем zlib.crc32 для эффективности
        with open(filepath, 'rb') as f:
            while True:
                chunk = f.read(8192)
                if not chunk:
                    break
                crc = zlib.crc32(chunk, crc)
                # if show_progress: ... 
        return crc & 0xFFFFFFFF
    else:
        # Для CRC32C используем табличный алгоритм
        import crcmod
        crc32c_func = crcmod.predefined.Crc('crc-32c')
        with open(filepath, 'rb') as f:
            while True:
                chunk = f.read(8192)
                if not chunk:
                    break
                crc32c_func.update(chunk)
        return crc32c_func.crcValue

def format_hash(hash_val, fmt='hex'):
    """Форматирует хэш в указанном формате."""
    if fmt == 'hex':
        return f"0x{hash_val:08X}"
    elif fmt == 'dec':
        return str(hash_val)
    elif fmt == 'bin':
        return f"{hash_val:032b}"
    return f"0x{hash_val:08X}"

def main():
    parser = argparse.ArgumentParser(description='CRC32 Hash Generator')
    parser.add_argument('inputs', nargs='*', help='Строки или файлы для обработки')
    parser.add_argument('--crc32c', action='store_true', help='Использовать CRC32C')
    parser.add_argument('--check', help='Сравнить с указанным хэшем (HEX)')
    parser.add_argument('--dec', action='store_true', help='Вывод в десятичном формате')
    parser.add_argument('--bin', action='store_true', help='Вывод в бинарном формате')
    parser.add_argument('--progress', action='store_true', help='Показывать прогресс для файлов')
    args = parser.parse_args()

    fmt = 'hex'
    if args.dec:
        fmt = 'dec'
    elif args.bin:
        fmt = 'bin'

    inputs = args.inputs
    if not inputs and not sys.stdin.isatty():
        # Чтение из stdin
        data = sys.stdin.buffer.read()
        hash_val = crc32_data(data, args.crc32c)
        print(format_hash(hash_val, fmt))
        return
    if not inputs:
        print("Не указаны данные. Используйте: crc32 <строка или файл>", file=sys.stderr)
        sys.exit(1)

    for item in inputs:
        # Проверяем, является ли аргумент файлом
        if os.path.isfile(item):
            hash_val = crc32_file(item, args.crc32c, args.progress)
            output = f"{item}: {format_hash(hash_val, fmt)}"
            if args.check:
                expected = int(args.check, 16) if args.check.startswith('0x') else int(args.check, 16)
                status = "✅ OK" if hash_val == expected else "❌ FAIL"
                output += f" (check: {status})"
            print(output)
        else:
            # Обрабатываем как строку
            data = item.encode('utf-8')
            hash_val = crc32_data(data, args.crc32c)
            output = format_hash(hash_val, fmt)
            if args.check:
                expected = int(args.check, 16) if args.check.startswith('0x') else int(args.check, 16)
                status = "✅ OK" if hash_val == expected else "❌ FAIL"
                output += f" (check: {status})"
            print(output)

if __name__ == '__main__':
    main()

File: crc32.rs
// crc32.rs
use std::fs::File;
use std::io::{self, Read, BufReader};
use std::path::Path;
use std::env;
use clap::{Arg, Command};
use crc::{Crc, Algorithm};

const CRC32_IEEE: Algorithm<u32> = Algorithm {
    width: 32,
    poly: 0x04C11DB7,
    init: 0xFFFFFFFF,
    refin: true,
    refout: true,
    xorout: 0xFFFFFFFF,
    check: 0xCBF43926,
    residue: 0xDEBB20E3,
};

const CRC32C: Algorithm<u32> = Algorithm {
    width: 32,
    poly: 0x1EDC6F41,
    init: 0xFFFFFFFF,
    refin: true,
    refout: true,
    xorout: 0xFFFFFFFF,
    check: 0xE3069283,
    residue: 0x48674BC7,
};

fn compute_crc32(data: &[u8], use_crc32c: bool) -> u32 {
    if use_crc32c {
        let crc = Crc::<u32>::new(&CRC32C);
        let mut digest = crc.digest();
        digest.update(data);
        digest.finalize()
    } else {
        let crc = Crc::<u32>::new(&CRC32_IEEE);
        let mut digest = crc.digest();
        digest.update(data);
        digest.finalize()
    }
}

fn compute_file_crc32(filename: &str, use_crc32c: bool) -> Result<u32, io::Error> {
    let file = File::open(filename)?;
    let mut reader = BufReader::new(file);
    let mut buffer = [0; 8192];
    if use_crc32c {
        let crc = Crc::<u32>::new(&CRC32C);
        let mut digest = crc.digest();
        loop {
            let n = reader.read(&mut buffer)?;
            if n == 0 { break; }
            digest.update(&buffer[..n]);
        }
        Ok(digest.finalize())
    } else {
        let crc = Crc::<u32>::new(&CRC32_IEEE);
        let mut digest = crc.digest();
        loop {
            let n = reader.read(&mut buffer)?;
            if n == 0 { break; }
            digest.update(&buffer[..n]);
        }
        Ok(digest.finalize())
    }
}

fn format_hash(val: u32, format: &str) -> String {
    match format {
        "hex" => format!("0x{:08X}", val),
        "dec" => format!("{}", val),
        "bin" => format!("{:032b}", val),
        _ => format!("0x{:08X}", val),
    }
}

fn main() {
    let matches = Command::new("crc32")
        .version("1.0")
        .about("CRC32 Hash Generator")
        .arg(Arg::new("inputs").multiple_occurrences(true).help("Строки или файлы"))
        .arg(Arg::new("crc32c").long("crc32c").help("Использовать CRC32C"))
        .arg(Arg::new("check").long("check").takes_value(true).help("Сравнить с хэшем (HEX)"))
        .arg(Arg::new("dec").long("dec").help("Вывод в десятичном формате"))
        .arg(Arg::new("bin").long("bin").help("Вывод в бинарном формате"))
        .arg(Arg::new("progress").long("progress").help("Показывать прогресс"))
        .get_matches();

    let use_crc32c = matches.is_present("crc32c");
    let check = matches.value_of("check").map(|s| u32::from_str_radix(s.trim_start_matches("0x"), 16));
    let format = if matches.is_present("dec") { "dec" } else if matches.is_present("bin") { "bin" } else { "hex" };
    let inputs: Vec<&str> = matches.get_many("inputs").unwrap_or_default().map(|s| *s).collect();

    if inputs.is_empty() {
        // Проверяем stdin
        let mut buffer = Vec::new();
        if let Ok(n) = io::stdin().read_to_end(&mut buffer) {
            if n > 0 {
                let hash_val = compute_crc32(&buffer, use_crc32c);
                println!("{}", format_hash(hash_val, format));
                return;
            }
        }
        eprintln!("Не указаны данные. Используйте: crc32 <строка или файл>");
        std::process::exit(1);
    }

    for item in inputs {
        let path = Path::new(item);
        if path.exists() && !path.is_dir() {
            match compute_file_crc32(item, use_crc32c) {
                Ok(hash_val) => {
                    let mut output = format!("{}: {}", item, format_hash(hash_val, format));
                    if let Some(Ok(expected)) = check {
                        let status = if expected == hash_val { "✅ OK" } else { "❌ FAIL" };
                        output.push_str(&format!(" (check: {})", status));
                    }
                    println!("{}", output);
                }
                Err(e) => eprintln!("Ошибка чтения {}: {}", item, e),
            }
        } else {
            let data = item.as_bytes();
            let hash_val = compute_crc32(data, use_crc32c);
            let mut output = format_hash(hash_val, format);
            if let Some(Ok(expected)) = check {
                let status = if expected == hash_val { "✅ OK" } else { "❌ FAIL" };
                output.push_str(&format!(" (check: {})", status));
            }
            println!("{}", output);
        }
    }
}
